File: app/src/main/java/com/example/shercofaqapp/viewmodel/GarageFragmentViewModel.java
package com.example.shercofaqapp.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.shercofaqapp.model.repositories.BikeFireBaseRepository;
import com.example.shercofaqapp.model.repositories.BikeRepository;
import com.example.shercofaqapp.model.Bike;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.ktx.Firebase;

import java.util.List;
import java.util.Objects;

public class GarageFragmentViewModel extends AndroidViewModel {

    BikeRepository bikeRepository;
    BikeFireBaseRepository bikeFireBaseRepository;
    private LiveData<List<Bike>> bikes;
    private LiveData<List<Bike>> bikesWithinFirebase;

    public GarageFragmentViewModel(@NonNull Application application) {
        super(application);

        bikeRepository = new BikeRepository(application);
        bikeFireBaseRepository = new BikeFireBaseRepository();
    }

    public LiveData<List<Bike>> getBikes() {
        bikes = bikeRepository.getBikes();
        return bikes;
    }

    public LiveData<List<Bike>> getBikesFromFirebase() {
        String userId = Objects.requireNonNull(FirebaseAuth.getInstance().getCurrentUser()).getUid();

        bikeFireBaseRepository.readBikes(userId, (MutableLiveData<List<Bike>>) bikesWithinFirebase);
        return bikesWithinFirebase;
    }

    public void addNewBike(Bike bike, String userId, String key) {
        bikeRepository.insertBike(bike);
        bikeFireBaseRepository.createBike(bike, userId, key);
    }

    public void deleteBike(Bike bike) {
        bikeRepository.deleteBike(bike);
    }

    public void updateBike(Bike bike, String userId, String bikeKey) {
        bikeRepository.updateBike(bike);
        bikeFireBaseRepository.updateBike(bike, userId, bikeKey);
    }

    public Bike getBike(long id) {
        return bikeRepository.getBike(id);
    }

}

File: app/src/main/java/com/example/shercofaqapp/viewmodel/GarageFragmentViewModel.kt
package com.example.shercofaqapp.viewmodel

import android.net.Uri
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.shercofaqapp.model.Bike
import com.example.shercofaqapp.model.repositories.BikeRepository
import com.example.shercofaqapp.utils.BIKES_NODE
import com.example.shercofaqapp.utils.CURRENT_UID
import com.example.shercofaqapp.utils.REF_STORAGE_ROOT
import com.example.shercofaqapp.utils.USERS_NODE
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class GarageFragmentViewModel
@Inject
constructor(
    private val bikeRepository: BikeRepository
) : ViewModel() {

    private var _bikeList = MutableLiveData<List<Bike>>()
    val bikeList : LiveData<List<Bike>>
        get() = _bikeList

    private var _bikeImageUrl = MutableLiveData<String>()
    val bikeImageUrl: LiveData<String>
        get() = _bikeImageUrl

    fun createBike(bike: Bike, bikeFirebaseKey: String) {
        bikeRepository.createBike(bike, bikeFirebaseKey)
    }

    fun getBikes() {
        bikeRepository.readBikes(_bikeList)
    }

    fun updateBike(bike: Bike, bikeKey: String) {
        bikeRepository.updateBike(bike, bikeKey)
    }

    fun deleteBike(bike: Bike) {
        bikeRepository.deleteBike(bike.bikeFirebaseKey!!)
    }

    fun getBikeImageUrl(bikeFirebaseKey: String, bikeImageUri: Uri) {
        val path = REF_STORAGE_ROOT
            .child(USERS_NODE)
            .child(CURRENT_UID)
            .child(BIKES_NODE)
            .child(bikeFirebaseKey)

        path.putFile(bikeImageUri).addOnCompleteListener { putFileTask ->
                if (putFileTask.isSuccessful) {
                    path.downloadUrl.addOnCompleteListener { downloadUrlTask ->
                        _bikeImageUrl.value = downloadUrlTask.result.toString()
                    }
                }
            }
    }

}